<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://www.jkobject.com/scdataloader/datamodule/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>datamodule - scdataloader</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "datamodule";
        var mkdocs_page_input_path = "datamodule.md";
        var mkdocs_page_url = "/scdataloader/datamodule/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scdataloader
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Example notebooks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../notebooks/1_download_and_preprocess/">one-off preparation and download of the database (optional)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../notebooks/2_create_dataloader/">Create the dataloader</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">documentation</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../dataset/">dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../preprocess/">preprocess</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../utils/">utils</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">datamodule</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#scdataloader.datamodule.DataModule">DataModule</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.decoders">decoders</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.genes">genes</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.labels_hierarchy">labels_hierarchy</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.predict_dataloader">predict_dataloader</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.setup">setup</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.test_dataloader">test_dataloader</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.train_dataloader">train_dataloader</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scdataloader.datamodule.DataModule.val_dataloader">val_dataloader</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../collator/">collator</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scdataloader</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">documentation</li>
      <li class="breadcrumb-item active">datamodule</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation-for-datamodule">Documentation for <code>DataModule</code></h1>


<div class="doc doc-object doc-class">



<h2 id="scdataloader.datamodule.DataModule" class="doc doc-heading">
              <code class="highlight language-python"><span class="n">scdataloader</span><span class="o">.</span><span class="n">datamodule</span><span class="o">.</span><span class="n">DataModule</span></code>

</h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="lightning.LightningDataModule">LightningDataModule</span></code></p>



        <p>PyTorch Lightning DataModule for loading single-cell data from a LaminDB Collection.</p>
<p>This DataModule provides train/val/test dataloaders with configurable sampling strategies.
It combines MappedCollection, Dataset, and Collator to create efficient data pipelines
for training single-cell foundation models.</p>
<p>The training dataloader uses weighted random sampling based on class frequencies,
validation uses random sampling, and test uses sequential sampling on held-out datasets.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>collection_name</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Key of the LaminDB Collection to load.</p>
              </div>
            </li>
            <li>
              <b><code>clss_to_weight</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>[&#39;organism_ontology_term_id&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>Label columns to use for weighted sampling
in the training dataloader. Supports "nnz" for weighting by number of
non-zero genes. Defaults to ["organism_ontology_term_id"].</p>
              </div>
            </li>
            <li>
              <b><code>weight_scaler</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>10</code>
)
              –
              <div class="doc-md-description">
                <p>Controls balance between rare and common classes.
Higher values lead to more uniform sampling across classes. Set to 0 to
disable weighted sampling. Defaults to 10.</p>
              </div>
            </li>
            <li>
              <b><code>n_samples_per_epoch</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>2000000</code>
)
              –
              <div class="doc-md-description">
                <p>Number of samples to draw per training epoch.
Defaults to 2,000,000.</p>
              </div>
            </li>
            <li>
              <b><code>validation_split</code></b>
                  (<code><span title="float">float</span> | <span title="int">int</span></code>, default:
                      <code>0.2</code>
)
              –
              <div class="doc-md-description">
                <p>Proportion (float) or absolute number (int)
of samples for validation. Defaults to 0.2.</p>
              </div>
            </li>
            <li>
              <b><code>test_split</code></b>
                  (<code><span title="float">float</span> | <span title="int">int</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Proportion (float) or absolute number (int)
of samples for testing. Uses entire datasets as test sets, rounding to
nearest dataset boundary. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>use_default_col</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to use the default Collator for batch
preparation. If False, no collate_fn is applied. Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>clss_to_predict</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>[&#39;organism_ontology_term_id&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>Observation columns to encode as prediction
targets. Must include "organism_ontology_term_id". Defaults to
["organism_ontology_term_id"].</p>
              </div>
            </li>
            <li>
              <b><code>hierarchical_clss</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>[]</code>
)
              –
              <div class="doc-md-description">
                <p>Observation columns with hierarchical
ontology structure to be processed. Defaults to [].</p>
              </div>
            </li>
            <li>
              <b><code>how</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;random expr&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Gene selection strategy passed to Collator. One of
"most expr", "random expr", "all", "some". Defaults to "random expr".</p>
              </div>
            </li>
            <li>
              <b><code>organism_col</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;organism_ontology_term_id&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Column name for organism ontology term ID.
Defaults to "organism_ontology_term_id".</p>
              </div>
            </li>
            <li>
              <b><code>max_len</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>1000</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum number of genes per sample passed to Collator.
Defaults to 1000.</p>
              </div>
            </li>
            <li>
              <b><code>replacement</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to sample with replacement in training.
Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>gene_subset</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>List of genes to restrict the dataset to.
Useful when model only supports specific genes. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>tp_name</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Column name for time point or heat diffusion values.
Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>assays_to_drop</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>[&#39;EFO:0030007&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>List of assay ontology term IDs to exclude
from training. Defaults to ["EFO:0030007"] (ATAC-seq).</p>
              </div>
            </li>
            <li>
              <b><code>metacell_mode</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.0</code>
)
              –
              <div class="doc-md-description">
                <p>Probability of using metacell aggregation mode.
Cannot be used with get_knn_cells. Defaults to 0.0.</p>
              </div>
            </li>
            <li>
              <b><code>get_knn_cells</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to include k-nearest neighbor cell
expression data. Cannot be used with metacell_mode. Defaults to False.</p>
              </div>
            </li>
            <li>
              <b><code>store_location</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Directory path to cache sampler indices and
labels for faster subsequent loading. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>force_recompute_indices</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Force recomputation of cached indices
even if they exist. Defaults to False.</p>
              </div>
            </li>
            <li>
              <b><code>sampler_workers</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Number of parallel workers for building sampler
indices. Auto-determined based on available CPUs if None. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>sampler_chunk_size</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Chunk size for parallel sampler processing.
Auto-determined based on available memory if None. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>organisms</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>List of organisms to include. If None, uses
all organisms in the dataset. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>genedf</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Gene information DataFrame. If None, loaded
automatically. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>n_bins</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Number of bins for expression discretization. 0 means
no binning. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>curiculum</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Curriculum learning parameter. If &gt; 0, gradually
increases sampling weight balance over epochs. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>start_at</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Starting index for resuming inference. Requires same
number of GPUs as previous run. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>**kwargs</code></b>
              –
              <div class="doc-md-description">
                <p>Additional arguments passed to PyTorch DataLoader (e.g., batch_size,
num_workers, pin_memory).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Attributes:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>dataset</code></b>
                  (<code><a class="autorefs autorefs-internal" title="scdataloader.data.Dataset

  
      dataclass
   (scdataloader.data.Dataset)" href="../dataset/#scdataloader.data.Dataset">Dataset</a></code>)
              –
              <div class="doc-md-description">
                <p>The underlying Dataset instance.</p>
              </div>
            </li>
            <li>
              <b><code>classes</code></b>
                  (<code><span title="dict">dict</span>[<span title="str">str</span>, <span title="int">int</span>]</code>)
              –
              <div class="doc-md-description">
                <p>Mapping from class names to number of categories.</p>
              </div>
            </li>
            <li>
              <b><code>train_labels</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Label array for weighted sampling.</p>
              </div>
            </li>
            <li>
              <b><code>idx_full</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Indices for training samples.</p>
              </div>
            </li>
            <li>
              <b><code>valid_idx</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Indices for validation samples.</p>
              </div>
            </li>
            <li>
              <b><code>test_idx</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Indices for test samples.</p>
              </div>
            </li>
            <li>
              <b><code>test_datasets</code></b>
                  (<code><span title="typing.List">List</span>[<span title="str">str</span>]</code>)
              –
              <div class="doc-md-description">
                <p>Paths to datasets used for testing.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If "organism_ontology_term_id" not in clss_to_predict.</p>
              </div>
            </li>
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If both metacell_mode &gt; 0 and get_knn_cells are True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>dm = DataModule(
...     collection_name="my_collection",
...     batch_size=32,
...     num_workers=4,
...     max_len=2000,
... )
dm.setup()
train_loader = dm.train_dataloader()</p>
</blockquote>
</blockquote>
</blockquote>
</details>










<p><span class="doc-section-title">Methods:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="predict_dataloader (scdataloader.datamodule.DataModule.predict_dataloader)" href="#scdataloader.datamodule.DataModule.predict_dataloader">predict_dataloader</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Create a DataLoader for prediction over all training data.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="setup (scdataloader.datamodule.DataModule.setup)" href="#scdataloader.datamodule.DataModule.setup">setup</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Prepare data splits for training, validation, and testing.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="test_dataloader (scdataloader.datamodule.DataModule.test_dataloader)" href="#scdataloader.datamodule.DataModule.test_dataloader">test_dataloader</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Create the test DataLoader with sequential sampling.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="train_dataloader (scdataloader.datamodule.DataModule.train_dataloader)" href="#scdataloader.datamodule.DataModule.train_dataloader">train_dataloader</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Create the training DataLoader with weighted random sampling.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="val_dataloader (scdataloader.datamodule.DataModule.val_dataloader)" href="#scdataloader.datamodule.DataModule.val_dataloader">val_dataloader</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Create the validation DataLoader.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>




<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Attributes:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>decoders</code></b>
              –
              <div class="doc-md-description">
                <p>decoders the decoders for any labels that would have been encoded</p>
              </div>
            </li>
            <li>
              <b><code>genes</code></b>
              –
              <div class="doc-md-description">
                <p>genes the genes used in this datamodule</p>
              </div>
            </li>
            <li>
              <b><code>labels_hierarchy</code></b>
              –
              <div class="doc-md-description">
                <p>labels_hierarchy the hierarchy of labels for any cls that would have a hierarchy</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
                  <details class="quote">
                    <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">collection_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">clss_to_weight</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">],</span>
    <span class="n">weight_scaler</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">n_samples_per_epoch</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2_000_000</span><span class="p">,</span>
    <span class="n">validation_split</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">test_split</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">use_default_col</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="c1"># this is for the mappedCollection</span>
    <span class="n">clss_to_predict</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">],</span>
    <span class="n">hierarchical_clss</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="c1"># this is for the collator</span>
    <span class="n">how</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;random expr&quot;</span><span class="p">,</span>
    <span class="n">organism_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">,</span>
    <span class="n">max_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">replacement</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">gene_subset</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tp_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># &quot;heat_diff&quot;</span>
    <span class="n">assays_to_drop</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="c1"># &quot;EFO:0008853&quot;, #patch seq</span>
        <span class="c1"># &quot;EFO:0010961&quot;, # visium</span>
        <span class="s2">&quot;EFO:0030007&quot;</span><span class="p">,</span>  <span class="c1"># ATACseq</span>
        <span class="c1"># &quot;EFO:0030062&quot;, # slide-seq</span>
    <span class="p">],</span>
    <span class="n">metacell_mode</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="n">get_knn_cells</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">store_location</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">force_recompute_indices</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">sampler_workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sampler_chunk_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">organisms</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">genedf</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">n_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">curiculum</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">start_at</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    PyTorch Lightning DataModule for loading single-cell data from a LaminDB Collection.</span>

<span class="sd">    This DataModule provides train/val/test dataloaders with configurable sampling strategies.</span>
<span class="sd">    It combines MappedCollection, Dataset, and Collator to create efficient data pipelines</span>
<span class="sd">    for training single-cell foundation models.</span>

<span class="sd">    The training dataloader uses weighted random sampling based on class frequencies,</span>
<span class="sd">    validation uses random sampling, and test uses sequential sampling on held-out datasets.</span>

<span class="sd">    Args:</span>
<span class="sd">        collection_name (str): Key of the LaminDB Collection to load.</span>
<span class="sd">        clss_to_weight (List[str], optional): Label columns to use for weighted sampling</span>
<span class="sd">            in the training dataloader. Supports &quot;nnz&quot; for weighting by number of</span>
<span class="sd">            non-zero genes. Defaults to [&quot;organism_ontology_term_id&quot;].</span>
<span class="sd">        weight_scaler (int, optional): Controls balance between rare and common classes.</span>
<span class="sd">            Higher values lead to more uniform sampling across classes. Set to 0 to</span>
<span class="sd">            disable weighted sampling. Defaults to 10.</span>
<span class="sd">        n_samples_per_epoch (int, optional): Number of samples to draw per training epoch.</span>
<span class="sd">            Defaults to 2,000,000.</span>
<span class="sd">        validation_split (float | int, optional): Proportion (float) or absolute number (int)</span>
<span class="sd">            of samples for validation. Defaults to 0.2.</span>
<span class="sd">        test_split (float | int, optional): Proportion (float) or absolute number (int)</span>
<span class="sd">            of samples for testing. Uses entire datasets as test sets, rounding to</span>
<span class="sd">            nearest dataset boundary. Defaults to 0.</span>
<span class="sd">        use_default_col (bool, optional): Whether to use the default Collator for batch</span>
<span class="sd">            preparation. If False, no collate_fn is applied. Defaults to True.</span>
<span class="sd">        clss_to_predict (List[str], optional): Observation columns to encode as prediction</span>
<span class="sd">            targets. Must include &quot;organism_ontology_term_id&quot;. Defaults to</span>
<span class="sd">            [&quot;organism_ontology_term_id&quot;].</span>
<span class="sd">        hierarchical_clss (List[str], optional): Observation columns with hierarchical</span>
<span class="sd">            ontology structure to be processed. Defaults to [].</span>
<span class="sd">        how (str, optional): Gene selection strategy passed to Collator. One of</span>
<span class="sd">            &quot;most expr&quot;, &quot;random expr&quot;, &quot;all&quot;, &quot;some&quot;. Defaults to &quot;random expr&quot;.</span>
<span class="sd">        organism_col (str, optional): Column name for organism ontology term ID.</span>
<span class="sd">            Defaults to &quot;organism_ontology_term_id&quot;.</span>
<span class="sd">        max_len (int, optional): Maximum number of genes per sample passed to Collator.</span>
<span class="sd">            Defaults to 1000.</span>
<span class="sd">        replacement (bool, optional): Whether to sample with replacement in training.</span>
<span class="sd">            Defaults to True.</span>
<span class="sd">        gene_subset (List[str], optional): List of genes to restrict the dataset to.</span>
<span class="sd">            Useful when model only supports specific genes. Defaults to None.</span>
<span class="sd">        tp_name (str, optional): Column name for time point or heat diffusion values.</span>
<span class="sd">            Defaults to None.</span>
<span class="sd">        assays_to_drop (List[str], optional): List of assay ontology term IDs to exclude</span>
<span class="sd">            from training. Defaults to [&quot;EFO:0030007&quot;] (ATAC-seq).</span>
<span class="sd">        metacell_mode (float, optional): Probability of using metacell aggregation mode.</span>
<span class="sd">            Cannot be used with get_knn_cells. Defaults to 0.0.</span>
<span class="sd">        get_knn_cells (bool, optional): Whether to include k-nearest neighbor cell</span>
<span class="sd">            expression data. Cannot be used with metacell_mode. Defaults to False.</span>
<span class="sd">        store_location (str, optional): Directory path to cache sampler indices and</span>
<span class="sd">            labels for faster subsequent loading. Defaults to None.</span>
<span class="sd">        force_recompute_indices (bool, optional): Force recomputation of cached indices</span>
<span class="sd">            even if they exist. Defaults to False.</span>
<span class="sd">        sampler_workers (int, optional): Number of parallel workers for building sampler</span>
<span class="sd">            indices. Auto-determined based on available CPUs if None. Defaults to None.</span>
<span class="sd">        sampler_chunk_size (int, optional): Chunk size for parallel sampler processing.</span>
<span class="sd">            Auto-determined based on available memory if None. Defaults to None.</span>
<span class="sd">        organisms (List[str], optional): List of organisms to include. If None, uses</span>
<span class="sd">            all organisms in the dataset. Defaults to None.</span>
<span class="sd">        genedf (pd.DataFrame, optional): Gene information DataFrame. If None, loaded</span>
<span class="sd">            automatically. Defaults to None.</span>
<span class="sd">        n_bins (int, optional): Number of bins for expression discretization. 0 means</span>
<span class="sd">            no binning. Defaults to 0.</span>
<span class="sd">        curiculum (int, optional): Curriculum learning parameter. If &gt; 0, gradually</span>
<span class="sd">            increases sampling weight balance over epochs. Defaults to 0.</span>
<span class="sd">        start_at (int, optional): Starting index for resuming inference. Requires same</span>
<span class="sd">            number of GPUs as previous run. Defaults to 0.</span>
<span class="sd">        **kwargs: Additional arguments passed to PyTorch DataLoader (e.g., batch_size,</span>
<span class="sd">            num_workers, pin_memory).</span>

<span class="sd">    Attributes:</span>
<span class="sd">        dataset (Dataset): The underlying Dataset instance.</span>
<span class="sd">        classes (dict[str, int]): Mapping from class names to number of categories.</span>
<span class="sd">        train_labels (np.ndarray): Label array for weighted sampling.</span>
<span class="sd">        idx_full (np.ndarray): Indices for training samples.</span>
<span class="sd">        valid_idx (np.ndarray): Indices for validation samples.</span>
<span class="sd">        test_idx (np.ndarray): Indices for test samples.</span>
<span class="sd">        test_datasets (List[str]): Paths to datasets used for testing.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If &quot;organism_ontology_term_id&quot; not in clss_to_predict.</span>
<span class="sd">        ValueError: If both metacell_mode &gt; 0 and get_knn_cells are True.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; dm = DataModule(</span>
<span class="sd">        ...     collection_name=&quot;my_collection&quot;,</span>
<span class="sd">        ...     batch_size=32,</span>
<span class="sd">        ...     num_workers=4,</span>
<span class="sd">        ...     max_len=2000,</span>
<span class="sd">        ... )</span>
<span class="sd">        &gt;&gt;&gt; dm.setup()</span>
<span class="sd">        &gt;&gt;&gt; train_loader = dm.train_dataloader()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;organism_ontology_term_id&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">clss_to_predict</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;need &#39;organism_ontology_term_id&#39; in the set of classes at least&quot;</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">metacell_mode</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">get_knn_cells</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;cannot use metacell mode and get_knn_cells at the same time&quot;</span>
        <span class="p">)</span>
    <span class="n">mdataset</span> <span class="o">=</span> <span class="n">Dataset</span><span class="p">(</span>
        <span class="n">ln</span><span class="o">.</span><span class="n">Collection</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">collection_name</span><span class="p">,</span> <span class="n">is_latest</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">(),</span>
        <span class="n">clss_to_predict</span><span class="o">=</span><span class="n">clss_to_predict</span><span class="p">,</span>
        <span class="n">hierarchical_clss</span><span class="o">=</span><span class="n">hierarchical_clss</span><span class="p">,</span>
        <span class="n">metacell_mode</span><span class="o">=</span><span class="n">metacell_mode</span><span class="p">,</span>
        <span class="n">get_knn_cells</span><span class="o">=</span><span class="n">get_knn_cells</span><span class="p">,</span>
        <span class="n">store_location</span><span class="o">=</span><span class="n">store_location</span><span class="p">,</span>
        <span class="n">force_recompute_indices</span><span class="o">=</span><span class="n">force_recompute_indices</span><span class="p">,</span>
        <span class="n">genedf</span><span class="o">=</span><span class="n">genedf</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># and location</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">metacell_mode</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="n">metacell_mode</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_pos</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">collection_name</span> <span class="o">=</span> <span class="n">collection_name</span>
    <span class="k">if</span> <span class="n">gene_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tokeep</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mdataset</span><span class="o">.</span><span class="n">genedf</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">gene_subset</span> <span class="o">=</span> <span class="p">[</span><span class="n">u</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">gene_subset</span> <span class="k">if</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tokeep</span><span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">classes</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">mdataset</span><span class="o">.</span><span class="n">class_topred</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="c1"># we might want not to order the genes by expression (or do it?)</span>
    <span class="c1"># we might want to not introduce zeros and</span>
    <span class="k">if</span> <span class="n">use_default_col</span><span class="p">:</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;collate_fn&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Collator</span><span class="p">(</span>
            <span class="n">organisms</span><span class="o">=</span><span class="n">mdataset</span><span class="o">.</span><span class="n">organisms</span> <span class="k">if</span> <span class="n">organisms</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">organisms</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="n">how</span><span class="p">,</span>
            <span class="n">valid_genes</span><span class="o">=</span><span class="n">gene_subset</span><span class="p">,</span>
            <span class="n">max_len</span><span class="o">=</span><span class="n">max_len</span><span class="p">,</span>
            <span class="n">org_to_id</span><span class="o">=</span><span class="n">mdataset</span><span class="o">.</span><span class="n">encoder</span><span class="p">[</span><span class="n">organism_col</span><span class="p">],</span>
            <span class="n">tp_name</span><span class="o">=</span><span class="n">tp_name</span><span class="p">,</span>
            <span class="n">organism_name</span><span class="o">=</span><span class="n">organism_col</span><span class="p">,</span>
            <span class="n">class_names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">classes</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
            <span class="n">genedf</span><span class="o">=</span><span class="n">genedf</span><span class="p">,</span>
            <span class="n">n_bins</span><span class="o">=</span><span class="n">n_bins</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_bins</span> <span class="o">=</span> <span class="n">n_bins</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">validation_split</span> <span class="o">=</span> <span class="n">validation_split</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">test_split</span> <span class="o">=</span> <span class="n">test_split</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span> <span class="o">=</span> <span class="n">mdataset</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">replacement</span> <span class="o">=</span> <span class="n">replacement</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span> <span class="o">=</span> <span class="n">kwargs</span>
    <span class="k">if</span> <span class="s2">&quot;sampler&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;sampler&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">assays_to_drop</span> <span class="o">=</span> <span class="n">assays_to_drop</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mdataset</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">weight_scaler</span> <span class="o">=</span> <span class="n">weight_scaler</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_samples_per_epoch</span> <span class="o">=</span> <span class="n">n_samples_per_epoch</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span> <span class="o">=</span> <span class="n">clss_to_weight</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">train_weights</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">train_labels</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">sampler_workers</span> <span class="o">=</span> <span class="n">sampler_workers</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">sampler_chunk_size</span> <span class="o">=</span> <span class="n">sampler_chunk_size</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">store_location</span> <span class="o">=</span> <span class="n">store_location</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">start_at</span> <span class="o">=</span> <span class="n">start_at</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_len</span> <span class="o">=</span> <span class="n">max_len</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">test_datasets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">force_recompute_indices</span> <span class="o">=</span> <span class="n">force_recompute_indices</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">curiculum</span> <span class="o">=</span> <span class="n">curiculum</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;finished init&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="scdataloader.datamodule.DataModule.decoders" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">decoders</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>decoders the decoders for any labels that would have been encoded</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>dict[str, dict[int, str]]</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="scdataloader.datamodule.DataModule.genes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">genes</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>genes the genes used in this datamodule</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>list</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="scdataloader.datamodule.DataModule.labels_hierarchy" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">labels_hierarchy</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>labels_hierarchy the hierarchy of labels for any cls that would have a hierarchy</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>dict[str, dict[str, str]]</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="scdataloader.datamodule.DataModule.predict_dataloader" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">predict_dataloader</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Create a DataLoader for prediction over all training data.</p>
<p>Uses RankShardSampler for distributed inference.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>DataLoader</code></b>              –
              <div class="doc-md-description">
                <p>Prediction DataLoader instance.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">predict_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a DataLoader for prediction over all training data.</span>

<span class="sd">    Uses RankShardSampler for distributed inference.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataLoader: Prediction DataLoader instance.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">subset</span> <span class="o">=</span> <span class="n">Subset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">DataLoader</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="p">,</span>
        <span class="n">sampler</span><span class="o">=</span><span class="n">RankShardSampler</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">subset</span><span class="p">),</span> <span class="n">start_at</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">start_at</span><span class="p">),</span>
        <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scdataloader.datamodule.DataModule.setup" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">setup</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Prepare data splits for training, validation, and testing.</p>
<p>This method shuffles the data, computes sample weights for weighted sampling,
removes samples from dropped assays, and creates train/val/test splits.
Test splits use entire datasets to ensure evaluation on unseen data sources.</p>
<p>Results can be cached to <code>store_location</code> for faster subsequent runs.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>stage</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Training stage ('fit', 'test', or None for both).
Currently not used but kept for Lightning compatibility. Defaults to None.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>List[str]: List of paths to test datasets.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<details class="note" open>
  <summary>Note</summary>
  <p>Must be called before using dataloaders. The train/val/test split is
deterministic when loading from cache.</p>
</details>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">setup</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stage</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare data splits for training, validation, and testing.</span>

<span class="sd">    This method shuffles the data, computes sample weights for weighted sampling,</span>
<span class="sd">    removes samples from dropped assays, and creates train/val/test splits.</span>
<span class="sd">    Test splits use entire datasets to ensure evaluation on unseen data sources.</span>

<span class="sd">    Results can be cached to `store_location` for faster subsequent runs.</span>

<span class="sd">    Args:</span>
<span class="sd">        stage (str, optional): Training stage (&#39;fit&#39;, &#39;test&#39;, or None for both).</span>
<span class="sd">            Currently not used but kept for Lightning compatibility. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: List of paths to test datasets.</span>

<span class="sd">    Note:</span>
<span class="sd">        Must be called before using dataloaders. The train/val/test split is</span>
<span class="sd">        deterministic when loading from cache.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;setting up the datamodule&quot;</span><span class="p">)</span>
    <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">store_location</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="ow">or</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;train_labels.npy&quot;</span><span class="p">))</span>
        <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">force_recompute_indices</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="s2">&quot;nnz&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">weight_scaler</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">get_merged_labels</span><span class="p">(</span>
                <span class="s2">&quot;nnz&quot;</span><span class="p">,</span> <span class="n">is_cat</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;nnz&quot;</span><span class="p">)</span>
            <span class="c1"># Sigmoid scaling with 2 parameters</span>
            <span class="n">midpoint</span> <span class="o">=</span> <span class="mi">2000</span>
            <span class="n">steepness</span> <span class="o">=</span> <span class="mf">0.003</span>
            <span class="c1"># Apply sigmoid transformation</span>
            <span class="c1"># sigmoid(x) = 1 / (1 + exp(-steepness * (x - midpoint)))</span>
            <span class="c1"># Then scale to [1, NNZ_SCALE] range</span>
            <span class="n">sigmoid_values</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">steepness</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="o">-</span> <span class="n">midpoint</span><span class="p">)))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="p">((</span><span class="n">NNZ_SCALE</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">sigmoid_values</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">weight_scaler</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">get_label_cats</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">validation_split</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">len_valid</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">validation_split</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">len_valid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">validation_split</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">test_split</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">len_test</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_split</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">len_test</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_split</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">len_test</span> <span class="o">+</span> <span class="n">len_valid</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span>
        <span class="p">),</span> <span class="s2">&quot;test set + valid set size is configured to be larger than entire dataset.&quot;</span>

        <span class="n">idx_full</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assays_to_drop</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">badloc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">get_merged_labels</span><span class="p">(</span>
                    <span class="s2">&quot;assay_ontology_term_id&quot;</span>
                <span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">assays_to_drop</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">idx_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))[</span><span class="o">~</span><span class="n">badloc</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">idx_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">len_test</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># this way we work on some never seen datasets</span>
            <span class="c1"># keeping at least one</span>
            <span class="n">len_test</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">len_test</span>
                <span class="k">if</span> <span class="n">len_test</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">n_obs_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">n_obs_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">cs</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">d_size</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">n_obs_list</span><span class="p">))</span>
            <span class="n">random</span><span class="o">.</span><span class="n">Random</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">d_size</span><span class="p">)</span>  <span class="c1"># always same order</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">d_size</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cs</span> <span class="o">+</span> <span class="n">c</span> <span class="o">&gt;</span> <span class="n">len_test</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">test_datasets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">mapped_dataset</span><span class="o">.</span><span class="n">path_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">path</span>
                    <span class="p">)</span>
                    <span class="n">cs</span> <span class="o">+=</span> <span class="n">c</span>
            <span class="n">len_test</span> <span class="o">=</span> <span class="n">cs</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="o">=</span> <span class="n">idx_full</span><span class="p">[:</span><span class="n">len_test</span><span class="p">]</span>
            <span class="n">idx_full</span> <span class="o">=</span> <span class="n">idx_full</span><span class="p">[</span><span class="n">len_test</span><span class="p">:]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">idx_full</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">len_valid</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="o">=</span> <span class="n">idx_full</span><span class="p">[:</span><span class="n">len_valid</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="c1"># store it for later</span>
            <span class="n">idx_full</span> <span class="o">=</span> <span class="n">idx_full</span><span class="p">[</span><span class="n">len_valid</span><span class="p">:]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">labels</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples</span><span class="p">),</span> <span class="n">idx_full</span><span class="p">)]</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="c1"># some labels will now not exist anymore as replaced by len(weights) - 1.</span>
        <span class="c1"># this means that the associated weights should be 0.</span>
        <span class="c1"># by doing np.bincount(labels)*weights this will be taken into account</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">train_labels</span> <span class="o">=</span> <span class="n">labels</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span> <span class="o">=</span> <span class="n">idx_full</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">store_location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;train_labels.npy&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">force_recompute_indices</span>
        <span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;nnz.npy&quot;</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span><span class="p">)</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;train_labels.npy&quot;</span><span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">train_labels</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;idx_full.npy&quot;</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;test_idx.npy&quot;</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;valid_idx.npy&quot;</span><span class="p">),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="n">listToFile</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">test_datasets</span><span class="p">,</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;test_datasets.txt&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nnz</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;nnz.npy&quot;</span><span class="p">),</span> <span class="n">mmap_mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;nnz.npy&quot;</span><span class="p">))</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">train_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;train_labels.npy&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;idx_full.npy&quot;</span><span class="p">),</span> <span class="n">mmap_mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;test_idx.npy&quot;</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;test_idx.npy&quot;</span><span class="p">))</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;valid_idx.npy&quot;</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;valid_idx.npy&quot;</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_datasets</span> <span class="o">=</span> <span class="n">fileToList</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span> <span class="s2">&quot;test_datasets.txt&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;loaded from store&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;done setup, took </span><span class="si">{</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start_time</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_datasets</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scdataloader.datamodule.DataModule.test_dataloader" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">test_dataloader</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Create the test DataLoader with sequential sampling.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>DataLoader | List: Test DataLoader, or empty list if no test split.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">test_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the test DataLoader with sequential sampling.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataLoader | List: Test DataLoader, or empty list if no test split.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">DataLoader</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="p">,</span> <span class="n">sampler</span><span class="o">=</span><span class="n">SequentialSampler</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span><span class="p">),</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_idx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="k">else</span> <span class="p">[]</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scdataloader.datamodule.DataModule.train_dataloader" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">train_dataloader</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Create the training DataLoader with weighted random sampling.</p>
<p>Uses LabelWeightedSampler for class-balanced sampling when weight_scaler &gt; 0
and clss_to_weight is specified. Otherwise uses RankShardSampler for
distributed training without weighting.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>**kwargs</code></b>
              –
              <div class="doc-md-description">
                <p>Additional arguments passed to DataLoader, overriding defaults.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>DataLoader</code></b>              –
              <div class="doc-md-description">
                <p>Training DataLoader instance.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If setup() has not been called.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">train_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the training DataLoader with weighted random sampling.</span>

<span class="sd">    Uses LabelWeightedSampler for class-balanced sampling when weight_scaler &gt; 0</span>
<span class="sd">    and clss_to_weight is specified. Otherwise uses RankShardSampler for</span>
<span class="sd">    distributed training without weighting.</span>

<span class="sd">    Args:</span>
<span class="sd">        **kwargs: Additional arguments passed to DataLoader, overriding defaults.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataLoader: Training DataLoader instance.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If setup() has not been called.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">clss_to_weight</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">weight_scaler</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Setting up the parallel train sampler...&quot;</span><span class="p">)</span>
            <span class="c1"># Create the optimized parallel sampler</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sampler_workers</span><span class="si">}</span><span class="s2"> workers for class indexing&quot;</span><span class="p">)</span>
            <span class="n">train_sampler</span> <span class="o">=</span> <span class="n">LabelWeightedSampler</span><span class="p">(</span>
                <span class="n">labels</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">train_labels</span><span class="p">,</span>
                <span class="n">weight_scaler</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">weight_scaler</span><span class="p">,</span>
                <span class="n">num_samples</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_samples_per_epoch</span><span class="p">),</span>
                <span class="n">element_weights</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">nnz</span><span class="p">,</span>
                <span class="n">replacement</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">replacement</span><span class="p">,</span>
                <span class="n">n_workers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sampler_workers</span><span class="p">,</span>
                <span class="n">chunk_size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sampler_chunk_size</span><span class="p">,</span>
                <span class="n">store_location</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">store_location</span><span class="p">,</span>
                <span class="n">force_recompute_indices</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">force_recompute_indices</span><span class="p">,</span>
                <span class="n">curiculum</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">curiculum</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; Have you run `datamodule.setup()`?&quot;</span><span class="p">)</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">Subset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">idx_full</span><span class="p">)</span>
        <span class="n">train_sampler</span> <span class="o">=</span> <span class="n">RankShardSampler</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">),</span> <span class="n">start_at</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">start_at</span><span class="p">)</span>
    <span class="n">current_loader_kwargs</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">current_loader_kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">DataLoader</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span> <span class="k">if</span> <span class="n">dataset</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">dataset</span><span class="p">,</span>
        <span class="n">sampler</span><span class="o">=</span><span class="n">train_sampler</span><span class="p">,</span>
        <span class="o">**</span><span class="n">current_loader_kwargs</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scdataloader.datamodule.DataModule.val_dataloader" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">val_dataloader</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Create the validation DataLoader.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>DataLoader | List: Validation DataLoader, or empty list if no validation split.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>scdataloader/datamodule.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">val_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the validation DataLoader.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataLoader | List: Validation DataLoader, or empty list if no validation split.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">DataLoader</span><span class="p">(</span>
            <span class="n">Subset</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span><span class="p">),</span>
            <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_idx</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="k">else</span> <span class="p">[]</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>


</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../utils/" class="btn btn-neutral float-left" title="utils"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../collator/" class="btn btn-neutral float-right" title="collator">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../utils/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../collator/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
